### PR TITLE
Fix #74: Add Checked.int8/uint8 and Nullable.int8/uint8/single/double

### LINKED ISSUE #74: Add Checked.int8/uint8, also Nullable.int8/uint8/single/double
This is the tracking issue for the "missing feature detail" request fixed [here](https://github.com/Microsoft/visualfsharp/pull/19) and with [UserVoice entry here](http://fslang.uservoice.com/forums/245727-f-language/suggestions/6995974-add-checked-int8-uint8-and-nullable-int8-uint8-sin)

F# defines "int8" and "uint8" as synonyms for "sbyte" and "byte", both as types and operators.  While reviewing the behaviour of "open Checked" (which brings new versions of operators into scope) I noticed that new versions of the "int8" and "uint8" operators are not defined in the "Checked" module. This is inconsistent, though a simple workaround is to use the checked "sbyte" and "byte" operators instead.

I also noticed that "Nullable.int8" and "Nullable.uint8", "Nullable.single" and "Nullable.double" are missing.

### PR DESCRIPTION
**Pull request by dsyme**
[Tracking UserVoice entry](http://fslang.uservoice.com/forums/245727-f-language/suggestions/6995974-add-checked-int8-uint8-and-nullable-int8-uint8-sin)

F# defines "int8" and "uint8" as synonyms for "sbyte" and "byte", both as types and operators

While reviewing the behaviour of "open Checked" (which brings new versions of operators into scope) I noticed that new versions of the "int8" and "uint8" operators are not defined in the "Checked" module.  This is inconsistent, though a simple workaround is to use the checked "sbyte" and "byte" operators instead.

I also noticed that "Nullable.int8" and "Nullable.uint8" are not defined.

All this should be fixed and the fix is simple.  I don't mind if it is in FSharp.Core 4.4.0.0 or later.

Cheers
Don

### CODE REVIEW

File: src/fsharp/FSharp.Core/fslib-extra-pervasives.fsi [hunk 1]
```diff
@@ -92,15 +92,31 @@ module ExtraTopLevelOperators =
     /// <summary>Converts the argument to byte.</summary>
     /// <remarks>This is a direct conversion for all 
     /// primitive numeric types. For strings, the input is converted using <c>Byte.Parse()</c> on strings and otherwise requires a <c>ToByte</c> method on the input type.</remarks>
-    [<CompiledName("ToByte")>]
-    val inline uint8       : value:^T -> byte       when ^T : (static member op_Explicit : ^T -> byte)       and default ^T : int        
+    [<CompiledName("ToUInt8")>]
+    val inline uint8       : value:^T -> uint8       when ^T : (static member op_Explicit : ^T -> byte)       and default ^T : int
```
Comment (mexx) [inline src/fsharp/FSharp.Core/fslib-extra-pervasives.fsi]: Should `static member op_Explicit : ^T -> byte` not be also `static member op_Explicit : ^T -> uint8` for consistency?

File: src/fsharp/FSharp.Core/fslib-extra-pervasives.fsi [hunk 2]
```diff
@@ -92,15 +92,31 @@ module ExtraTopLevelOperators =
     /// <summary>Converts the argument to byte.</summary>
     /// <remarks>This is a direct conversion for all 
     /// primitive numeric types. For strings, the input is converted using <c>Byte.Parse()</c> on strings and otherwise requires a <c>ToByte</c> method on the input type.</remarks>
-    [<CompiledName("ToByte")>]
-    val inline uint8       : value:^T -> byte       when ^T : (static member op_Explicit : ^T -> byte)       and default ^T : int        
+    [<CompiledName("ToUInt8")>]
```
Comment (mexx) [inline src/fsharp/FSharp.Core/fslib-extra-pervasives.fsi]: Do I get it right, this will be a breaking change for external usages?

Comment (dsyme) [inline src/fsharp/FSharp.Core/fslib-extra-pervasives.fsi]: I have reverted this change.  The compiled name should really always have been "ToUInt8", but you are correct, we won't change it now for backwards-compat reasons.

**Comment by dsyme** [comment]
I've updated the PR with 
* more testing (including for the added nullable ops)
* Nullable.double and Nullable.single, which wewre also missing.

I'll write up a mini-spec for this too.

File: src/fsharp/FSharp.Core/Linq.fsi [hunk 3]
```diff
@@ -109,6 +117,15 @@ namespace Microsoft.FSharp.Linq
         [<CompiledName("ToSByte")>]
         val inline sbyte      : value:Nullable< ^T > -> Nullable<sbyte>      when ^T : (static member op_Explicit : ^T -> sbyte)      and default ^T : int
         
+        
+        /// <summary>Converts the argument to signed byte. This is a direct conversion for all 
+        /// primitive numeric types. The operation requires an appropriate
+        /// static conversion method on the input type.</summary>
+        /// <param name="value">The input value.</param>
+        /// <returns>The converted sbyte</returns>
+        [<CompiledName("ToInt8")>]
```
Comment (latkin) [inline src/fsharp/FSharp.Core/Linq.fsi]: If new APIs are being added, you will need to update the surface area tests in FSharp.Core.Unittests.  Note that ATM this test is failing, as some required updates were missed during recent PRs. @KevinRansom I think is working on fixing it.

**Comment by latkin** [comment]
This looks like a combo mini-feature and cleanup/refactor. Do I understand correctly?

If so, I will direct you to our contribution guidelines, written by a wise contributor to this repo :smile: 
> Code cleanup should generally be made separate to other checkins where possible, or the parts that are cleanup should be labelled as such.

File: src/fsharp/FSharp.Core/Linq.fs [hunk 4]
```diff
@@ -302,30 +319,41 @@ module LeafExpressionConverter =
...
-    let (|ConvSByteQ|_|)       = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> sbyte x))
-    let (|ConvInt16Q|_|)       = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> int16 x))
-    let (|ConvInt32Q|_|)       = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> int32 x))
-    let (|ConvIntQ|_|)         = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> int x))
-    let (|ConvInt64Q|_|)       = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> int64 x))
-    let (|ConvByteQ|_|)        = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> byte x))
-    let (|ConvUInt16Q|_|)      = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> uint16 x))
-    let (|ConvUInt32Q|_|)      = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> uint32 x))
-    let (|ConvUInt64Q|_|)      = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> uint64 x))
+    let (|ConvCharQ|_|)        = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> Operators.char x))
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/Linq.fs]: Throughotu this file "Operators." has been included to give important code clarification as part of the bug fix as to which operator implementation is being called

**Comment by dsyme** [comment]
great, thanks

- Yes, this is a feature.

- I've reduced the code change to what's needed for a clean, clarifying bugfix

- I've added a UserVoice entry (since all features need one)

- I've updated the surface area tests

File: src/fsharp/FSharp.Core/Linq.fs [hunk 5]
```diff
@@ -77,53 +77,66 @@ module NullableOperators =
...
-        let inline int (x:Nullable< ^T >) = if x.HasValue then Nullable(int x.Value) else Nullable()
+        let inline int (x:Nullable< ^T >) = if x.HasValue then Nullable(Operators.int x.Value) else Nullable()
         
         [<CompiledName("ToEnum")>]
-        let inline enum (x:Nullable< int32 >) = if x.HasValue then Nullable(enum x.Value) else Nullable()
+        let inline enum (x:Nullable< int32 >) = if x.HasValue then Nullable(Operators.enum x.Value) else Nullable()
 
         [<CompiledName("ToInt32")>]
-        let inline int32 (x:Nullable<_>) = if x.HasValue then Nullable(int32 x.Value) else Nullable()
+        let inline int32 (x:Nullable<_>) = if x.HasValue then Nullable(Operators.int32 x.Value) else Nullable()
```
Comment (enricosada) [inline src/fsharp/FSharp.Core/Linq.fs]: why `Nullable.int16` parameter is `x: Nullable< ^T >` but `Nullable.int32` is `x: Nullable<_>` ?

Comment (dsyme) [inline src/fsharp/FSharp.Core/Linq.fs]: It doesn't matter either way. However I won't normalize it as part of this PR since it's unrelated cleanup.

Comment (enricosada) [inline src/fsharp/FSharp.Core/Linq.fs]: @dsyme, sure, i was curious to learn if there was a diff behaviour in this context or only code style. ty

**Comment by latkin** [comment]
Hmm, since this will alter the public surface area of FSharp.Core, it really needs to be done for 4.0, or else wait for the next version bump (might be a while). Seems innocuous enough... @KevinRansom what do you think, should we pull this for 4.0?

**Comment by KevinRansom** [comment]
Absolutely, good eyes on spotting the inconsistency.
